documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_moderation_metrics.md

```diff
@@ -0,0 +1,327 @@
+and our various responses over time. Similar queries can be used to create
+charts for other metrics. Ultimately a dashboard can be put together in
+CloudWatch, that tracks all relevant metrics.
+
+This dashboard will also be tracked in the infrastructure repo for backup and
+restoration purposes.
+
+### Implement deferred metrics
+
+Deferred metrics are those that are computed periodically or on-demand from the
+data in our database. These metrics can be computed using the Django ORM when
+triggered using a management command and their results can be cached.
+
+Here is how these deferred metrics can be calculated. Assume `Report`,
+`Decision` and `Media` here refer to image and audio models respectively. This
+example uses [package `tailslide`](https://github.com/ankane/tailslide) but a
+fairly basic aggregator could be implemented by us if needed.
+
+```python
+from tailslide import Percentile
+from django.db.models import Avg, Count, F
+from api.models import Report, Decision, Media
+from datetime import timedelta
+
+# 1. Report Accuracy
+total_reports = Report.objects.count()
+confirmed_reports = Report.objects.filter(decision__action='confirmed').count()
+report_accuracy = (confirmed_reports / total_reports) * 100 if total_reports else 0
+
+# 2. Report Duplication
+duplicate_reports = Report.objects.filter(decision__action='deduplicated').count()
+report_duplication = (duplicate_reports / total_reports) * 100 if total_reports else 0
+
+# 3. Most Reported Items
+most_reported_media = Report.objects.values('media_id').annotate(report_count=Count('id')).order_by('-report_count').first()
+most_reported_creator = Report.objects.values('media__creator').annotate(report_count=Count('id')).order_by('-report_count').first()
+most_reported_source = Report.objects.values('media__source').annotate(report_count=Count('id')).order_by('-report_count').first()
+
+# 4. Average and P99 Time to Resolution
+time_differences = Report.objects.annotate(
+  resolution_time=Avg(F('decision__created_on') - F('created_on'))
+).values_list('resolution_time', flat=True)
+
+average_resolution_time = sum(time_differences, timedelta()) / len(time_differences) if time_differences else timedelta()
+p99_resolution_time = Report.objects.annotate(
+  resolution_time=F('decision__created_on') - F('created_on')
+).aggregate(
+  percentile_99=Percentile('resolution_time', 0.99)
+)['percentile_99']
+```
+
+### Surface in Django Admin
+
+The admin site supports both adding
+[custom views](https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#adding-views-to-admin-sites)
+and
+[overriding/extending existing templates](https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#overriding-admin-templates).
+
+These approaches can be coupled together for a good experience in displaying
+these metrics to the moderators.
+
+The precise design of how these metrics will be rendered will depend on the UI
+of the moderation views as implemented when the previous IP is realised.
```

Comment:
Can you clarify how it will depend on the _UI_ of the moderation views?

If I was picking up an issue for this work, I'd have trouble envisioning what is planned here -- are you picturing a static template? Would we have custom filters, for example, for selecting results by provider/source/action type/date range?

Reply:
I agree with this. I should have dived a bit deeper into the UI specifics because I left them too vague. I have elaborated on this in bb3cbfb, hopefully it makes my vision of how the UI would be presented much clearer.